README.md
![GitHub Workflow Status (event)](https://img.shields.io/github/workflow/status/weiliang-ms/easyctl/Go?style=flat-square)
[![Go Report Card](https://goreportcard.com/badge/github.com/weiliang-ms/easyctl)](https://goreportcard.com/report/github.com/weiliang-ms/easyctl)
![GitHub go.mod Go version (subdirectory of monorepo)](https://img.shields.io/github/go-mod/go-version/weiliang-ms/easyctl?filename=go.mod&style=flat-square)
![GitHub commit activity](https://img.shields.io/github/commit-activity/w/weiliang-ms/easyctl?style=flat-square)
![GitHub all releases](https://img.shields.io/github/downloads/weiliang-ms/easyctl/total?style=flat-square)
[![codecov](https://codecov.io/gh/weiliang-ms/easyctl/branch/master/graph/badge.svg?token=7RGD5V5L9Y)](https://codecov.io/gh/weiliang-ms/easyctl)
![GitHub](https://img.shields.io/github/license/weiliang-ms/easyctl?style=flat-square)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fweiliang-ms%2Feasyctl.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fweiliang-ms%2Feasyctl?ref=badge_shield)

# easyctl

`easyctl`是一款基于`golang`轻量级运维工具集，详情使用请移步[使用文档](https://weiliang-ms.github.io/easyctl/)

[CHANGELOG](changelog/CHANGELOG.md)

兼容性：

- [x] 全部功能兼容`CentOS7`
- [ ] 部分功能兼容`CentOS6`

## 安装使用

### 编译安装最新版

> Go 版本不得低于 v1.16

> 需要设置代理时，可参照 https://goproxy.cn/

```shell
git clone https://github.com/weiliang-ms/easyctl.git
cd easyctl
go build -ldflags "-w -s" -o /usr/local/bin/easyctl
```

> [关于 -ldflags "-w -s"](https://stackoverflow.com/questions/44148449/how-to-check-whether-golang-binary-is-compiled-with-ldflags-s-w)

#### Mac 编译

Mac 下编译时会有如下提示：

```bash
$ go build
# github.com/containerd/cgroups
vendor/github.com/containerd/cgroups/memory.go:463:14: undefined: unix.Eventfd
vendor/github.com/containerd/cgroups/memory.go:463:30: undefined: unix.EFD_CLOEXEC
vendor/github.com/containerd/cgroups/utils.go:70:8: undefined: unix.CGROUP2_SUPER_MAGIC
vendor/github.com/containerd/cgroups/utils.go:77:18: undefined: unix.CGROUP2_SUPER_MAGIC
```

因为 Mac 下无法使用 containerd，可将 [启动指令集](https://weiliang-ms.github.io/easyctl/%E5%90%AF%E5%8A%A8%E6%8C%87%E4%BB%A4%E9%9B%86/%E6%B7%BB%E5%8A%A0%E9%85%8D%E9%A2%9D%E5%90%AF%E5%8A%A8%E7%A8%8B%E5%BA%8F.html) 相关内容 [注释](https://github.com/AlphaHinex/easyctl/compare/master...AlphaHinex:mac-build)，即可成功编译。

### 下载release版本

> [latest release](https://github.com/weiliang-ms/easyctl/releases)

- [Mac OS v0.7.13-alpha](https://github.com/weiliang-ms/easyctl/releases/download/v0.7.13-alpha/easyctl-v0.7.13-alpha-darwin-amd64.tar.gz)
```shell
tar zxvf easyctl-v0.7.13-alpha-darwin-amd64.tar.gz
cp easyctl /usr/local/bin
```
系统偏好设置 => 安全性与隐私 => 通用 => 仍然允许

- [linux-amd64](https://github.com/weiliang-ms/easyctl/releases/download/latest/easyctl-latest-linux-amd64.tar.gz)
```shell
sudo tar zxvf easyctl-latest-linux-amd64.tar.gz
sudo cp easyctl /usr/local/bin
```

> 下载[tag版本](https://github.com/weiliang-ms/easyctl/tags)

## 迭代计划

> 里程碑

- [x] [文档站点](https://weiliang-ms.github.io/easyctl/)
- [x] 集成`github workflow`
- [x] [集成codecov](https://app.codecov.io/gh/weiliang-ms/+)
- `v0.x.y-alpha`: 开发
- `v1.0.0-beta`: `bug`修复
- `v1.0.0-release`: 正式版本

> `v1`功能列表

- `add`
- `boot`
  - [x] [以配额的方式启动程序](https://weiliang-ms.github.io/easyctl/%E5%90%AF%E5%8A%A8%E6%8C%87%E4%BB%A4%E9%9B%86/%E6%B7%BB%E5%8A%A0%E9%85%8D%E9%A2%9D%E5%90%AF%E5%8A%A8%E7%A8%8B%E5%BA%8F.html)
- `clean`
  - [x] [清理多机redis](https://weiliang-ms.github.io/easyctl/%E6%B8%85%E7%90%86%E6%8C%87%E4%BB%A4%E9%9B%86/01%E5%A4%9A%E4%B8%BB%E6%9C%BAredis%E6%B8%85%E7%90%86.html)
  - [x] [清理主机dns配置](https://weiliang-ms.github.io/easyctl/%E6%B8%85%E7%90%86%E6%8C%87%E4%BB%A4%E9%9B%86/01%E5%A4%9A%E4%B8%BB%E6%9C%BAdns%E9%85%8D%E7%BD%AE%E6%B8%85%E7%90%86.html)
- `deny`
  - [x] [防火墙](https://weiliang-ms.github.io/easyctl/%E7%A6%81%E7%94%A8%E6%8C%87%E4%BB%A4%E9%9B%86/03%E4%B8%BB%E6%9C%BA%E7%A6%81%E7%94%A8%E9%98%B2%E7%81%AB%E5%A2%99.html)
  - [x] [ping](https://weiliang-ms.github.io/easyctl/%E7%A6%81%E7%94%A8%E6%8C%87%E4%BB%A4%E9%9B%86/01%E4%B8%BB%E6%9C%BA%E7%A6%81Ping.html)
  - [x] [selinux](https://weiliang-ms.github.io/easyctl/%E7%A6%81%E7%94%A8%E6%8C%87%E4%BB%A4%E9%9B%86/02%E4%B8%BB%E6%9C%BA%E7%A6%81%E7%94%A8selinux.html)
- `exec`
  - [x] [跨主机并行执行shell](https://weiliang-ms.github.io/easyctl/%E6%89%A7%E8%A1%8C%E6%8C%87%E4%BB%A4%E9%9B%86/01%E8%B7%A8%E4%B8%BB%E6%9C%BA%E5%B9%B6%E8%A1%8C%E6%89%A7%E8%A1%8Cshell.html)
  - [x] [多主机ping检测](https://weiliang-ms.github.io/easyctl/%E6%89%A7%E8%A1%8C%E6%8C%87%E4%BB%A4%E9%9B%86/02%E5%A4%9A%E6%9C%BAping.html)
- `export`
  - [x] [从harbor批量下载chart](https://weiliang-ms.github.io/easyctl/%E5%AF%BC%E5%87%BA%E6%8C%87%E4%BB%A4%E9%9B%86/01%E4%BB%8Eharbor%E5%BA%93%E4%B8%AD%E5%AF%BC%E5%87%BAchart.html)
  - [x] [从harbor导出镜像tag列表](https://weiliang-ms.github.io/easyctl/%E5%AF%BC%E5%87%BA%E6%8C%87%E4%BB%A4%E9%9B%86/02%E4%BB%8Eharbor%E5%BA%93%E4%B8%AD%E5%AF%BC%E5%87%BA%E9%95%9C%E5%83%8Ftag%E5%88%97%E8%A1%A8.html)
- `harden`
  - [x] [操作系统加固](https://weiliang-ms.github.io/easyctl/%E5%8A%A0%E5%9B%BA%E6%8C%87%E4%BB%A4%E9%9B%86/%E5%8A%A0%E5%9B%BA%E6%93%8D%E4%BD%9C%E7%B3%BB%E7%BB%9F.html)
- `install`
  - [x] [二进制方式安装docker](book/安装指令集/03安装docker.md)
  - [x] [单机redis](https://weiliang-ms.github.io/easyctl/%E5%AE%89%E8%A3%85%E6%8C%87%E4%BB%A4%E9%9B%86/01%E5%AE%89%E8%A3%85%E5%8D%95%E6%9C%BAredis.html)
  - [x] [redis集群](https://weiliang-ms.github.io/easyctl/%E5%AE%89%E8%A3%85%E6%8C%87%E4%BB%A4%E9%9B%86/02%E5%AE%89%E8%A3%85redis%E9%9B%86%E7%BE%A4.html)
- `set`
  - [x] [时区](https://weiliang-ms.github.io/easyctl/%E8%AE%BE%E7%BD%AE%E6%8C%87%E4%BB%A4%E9%9B%86/05%E5%A4%9A%E4%B8%BB%E6%9C%BA%E8%AE%BE%E7%BD%AE%E6%97%B6%E5%8C%BA.html)
  - [x] [dns](https://weiliang-ms.github.io/easyctl/%E8%AE%BE%E7%BD%AE%E6%8C%87%E4%BB%A4%E9%9B%86/06%E5%A4%9A%E4%B8%BB%E6%9C%BA%E9%85%8D%E7%BD%AEdns.html)
  - [x] [主机互信](https://weiliang-ms.github.io/easyctl/%E8%AE%BE%E7%BD%AE%E6%8C%87%E4%BB%A4%E9%9B%86/03%E5%A4%9A%E4%B8%BB%E6%9C%BA%E5%85%8D%E5%AF%86%E7%99%BB%E5%BD%95.html)
  - [x] [主机host解析](https://weiliang-ms.github.io/easyctl/%E8%AE%BE%E7%BD%AE%E6%8C%87%E4%BB%A4%E9%9B%86/01%E5%A4%9A%E4%B8%BB%E6%9C%BAhost%E8%A7%A3%E6%9E%90.html)
  - [x] [文件描述符数](https://weiliang-ms.github.io/easyctl/%E8%AE%BE%E7%BD%AE%E6%8C%87%E4%BB%A4%E9%9B%86/04%E5%A4%9A%E4%B8%BB%E6%9C%BA%E8%AE%BE%E7%BD%AE%E6%96%87%E4%BB%B6%E6%8F%8F%E8%BF%B0%E7%AC%A6.html)
  - [x] [修改root口令](https://weiliang-ms.github.io/easyctl/%E8%AE%BE%E7%BD%AE%E6%8C%87%E4%BB%A4%E9%9B%86/02%E5%A4%9A%E4%B8%BB%E6%9C%BA%E4%BF%AE%E6%94%B9root%E5%8F%A3%E4%BB%A4.html)
- `track`
  - [x] [日志tail](https://weiliang-ms.github.io/easyctl/%E8%BF%BD%E8%B8%AA%E6%8C%87%E4%BB%A4%E9%9B%86/01%E5%A4%9A%E4%B8%BB%E6%9C%BA%E6%97%A5%E5%BF%97%E5%AE%9E%E6%97%B6%E8%BF%BD%E8%B8%AA.html)

> `v2`功能预览

- `windows GUI`

## License
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fweiliang-ms%2Feasyctl.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fweiliang-ms%2Feasyctl?ref=badge_large)


changelog/CHANGELOG.md
- `v0.7.18-alpha`  `2021-12-31 15:07:00`
  - 调整本地安装`docker-ce`方式
- `v0.7.17-alpha`  `2021-12-29 08:28:00`
  - 添加`exec`指令集子指令`ping`
- `v0.7.16-alpha`  `2021-12-22 08:28:00`
  - 添加`exec`指令集子指令`shell`
- `v0.7.15-alpha`  `2021-11-05 11:28:00`
  - 添加`install`指令集子指令`docker-ce`
- `v0.7.14-alpha`  `2021-10-26 17:00:00`
  - 添加`boot`指令集子指令`app-with-cgroups`
- `v0.7.13-alpha`  `2021-10-20 17:00:00`
  - `install redis-cluster`指令添加端口取值范围设置
- `v0.7.12-alpha`  `2021-10-11 17:00:00`
  - 添加清理`harden`系统加固子指令
  - 重构执行器返回错误类型
- `v0.7.11-alpha`  `2021-10-10 19:48:00`
  - 添加清理`redis`指令
- `v0.7.10-alpha`  `2021-10-10 14:00:00`
  - 提交部分指令测试用例
  - 重构`cmd`包内指令调用函数返回值
  - 添加`redis`安装指令
- `v0.7.9-alpha`  `2021-10-09 21:32:00`
  - 提交部分指令测试用例
  - 重构`cmd`包内`cmd`指令调用函数
  - 集成`travis CI`
  - 集成`FOSSA license scan`
  - 添加`redis`集群安装指令

main.go
package main

import (
	// embed
	_ "embed"
	"fmt"
	"github.com/spf13/cobra"
	"github.com/weiliang-ms/easyctl/cmd/add"
	"github.com/weiliang-ms/easyctl/cmd/boot"
	"github.com/weiliang-ms/easyctl/cmd/clean"
	"github.com/weiliang-ms/easyctl/cmd/deny"
	"github.com/weiliang-ms/easyctl/cmd/exec"
	"github.com/weiliang-ms/easyctl/cmd/export"
	"github.com/weiliang-ms/easyctl/cmd/harden"
	"github.com/weiliang-ms/easyctl/cmd/install"
	"github.com/weiliang-ms/easyctl/cmd/scan"
	"github.com/weiliang-ms/easyctl/cmd/set"
	"github.com/weiliang-ms/easyctl/cmd/track"
	"github.com/weiliang-ms/easyctl/cmd/upgrade"
	"math/rand"
	"os"
	"time"
)

var (
	// GitTag git branch
	GitTag = "2000.01.01.release"
	// BuildTime 构建时间
	BuildTime = "2000-01-01T00:00:00+0800"
	// Debug 是否开启debug
	Debug bool
)

// RootCmd 根命令
var RootCmd = &cobra.Command{
	Use:   "easyctl",
	Short: "Easycf is a tool manage linux settings",
	Long: `A Fast and Flexible Static Site Generator built with
                love by spf13 and friends in Go.
                Complete documentation is available at http://hugo.spf13.com`,
	Args: cobra.ExactValidArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		// Do Stuff Here
	},
}

func init() {

	RootCmd.PersistentFlags().BoolVarP(&Debug, "debug", "d", false, "开启debug模式")
	subCmds := []*cobra.Command{
		add.RootCmd,
		boot.RootCmd,
		clean.RootCmd,
		deny.RootCmd,
		exec.RootCmd,
		scan.RootCmd,
		set.RootCmd,
		install.RootCmd,
		upgrade.Cmd,
		//cmd.InitTmplCmd,
		export.RootCmd,
		harden.RootCmd,
		track.RootCmd,
		versionCmd,
		completionCmd,
	}

	for _, cmd := range subCmds {
		RootCmd.AddCommand(cmd)
	}

}

func main() {

	rand.Seed(time.Now().UnixNano())

	if err := RootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

// 输出easyctl版本
var versionCmd = &cobra.Command{
	Use:   "version",
	Short: "Print the version number of easyctl",
	Long:  `All software has versions. This is easyctl's`,
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Printf("easyctl %s -- %s\n", GitTag, BuildTime)
		//fmt.Println("easyctl  v0.6.0 -- alpha -- 2021-06-22-09:58:00")
	},
}

var completionCmd = &cobra.Command{
	Use:   "completion [bash|zsh|fish|powershell]",
	Short: "Generate completion script",
	Long: `To load completions:

Bash:

$ source <(yourprogram completion bash)

# To load completions for each session, execute once:
Linux:
  $ yourprogram completion bash > /etc/bash_completion.d/yourprogram
MacOS:
  $ yourprogram completion bash > /usr/local/etc/bash_completion.d/yourprogram

Zsh:

# If shell completion is not already enabled in your environment you will need
# to enable it.  You can execute the following once:

$ echo "autoload -U compinit; compinit" >> ~/.zshrc

# To load completions for each session, execute once:
$ yourprogram completion zsh > "${fpath[1]}/_yourprogram"

# You will need to start a new shell for this setup to take effect.

Fish:

$ yourprogram completion fish | source

# To load completions for each session, execute once:
$ yourprogram completion fish > ~/.config/fish/completions/yourprogram.fish
`,
	DisableFlagsInUseLine: true,
	ValidArgs:             []string{"bash", "zsh", "fish", "powershell"},
	Args:                  cobra.ExactValidArgs(1),
	Hidden:                true,
	Run: func(cmd *cobra.Command, args []string) {
		switch args[0] {
		case "bash":
			cmd.Root().GenBashCompletion(os.Stdout)
		case "zsh":
			cmd.Root().GenZshCompletion(os.Stdout)
		case "fish":
			cmd.Root().GenFishCompletion(os.Stdout, true)
		case "powershell":
			cmd.Root().GenPowerShellCompletion(os.Stdout)
		}
	},
}
